src/target/riscv/riscv.c

```diff
@@ -700,6 +734,110 @@ static void log_trigger_request_info(struct trigger_request_info trig_info)
 			trig_info.tdata1, trig_info.tdata2, trig_info.tdata1_ignore_mask);
 };
 
+static struct tdata1_cache *create_new_tdata1_cache(struct list_head *tdata1_cache_head, riscv_reg_t tdata1)
+{
+
+	struct tdata1_cache *elem = (struct tdata1_cache *)calloc(1, sizeof(struct tdata1_cache));
+	elem->tdata1 = tdata1;
+	INIT_LIST_HEAD(&elem->tdata2_cache_head);
+	list_add_tail(&elem->elem_tdata1, tdata1_cache_head);
+	return elem;
+}
+
+static void create_new_tdata2_cache(struct list_head *tdata2_cache_head, riscv_reg_t tdata2)
+{
+
+	struct tdata2_cache * const elem = calloc(1, sizeof(struct tdata2_cache));
+	elem->tdata2 = tdata2;
+	list_add(&elem->elem_tdata2, tdata2_cache_head);
+}
+
+struct tdata2_cache *find_call_in_tdata2_cache(struct list_head *tdata2_cache_head, riscv_reg_t find_tdata2)
+{
+
+	struct tdata2_cache *elem_2;
+	list_for_each_entry(elem_2, tdata2_cache_head, elem_tdata2) {
+		if (elem_2->tdata2 == find_tdata2) {
+			return elem_2;
+		}
+	}
+	return NULL;
+}
+
+struct tdata1_cache *find_tdata1(struct list_head *tdata1_cache_head, riscv_reg_t find_tdata1)
+{
+
+	struct tdata1_cache *elem_1;
+	list_for_each_entry(elem_1, tdata1_cache_head, elem_tdata1) {
+		if (elem_1->tdata1 == find_tdata1)
+			return elem_1;
+	}
+	return NULL;
+}
+
+static void create_wp_trigger_cache(struct target *target)
+{
+
+	RISCV_INFO(r);
+
+	r->wp_triggers_cache = (struct list_head *)calloc(r->trigger_count,
+		sizeof(struct list_head));
```

Comment:
One cache per trigger makes sense. But then all you need is the contents of tdata1/tdata2 along with some bits to mark whether they are valid, right?

I completely do not understand why you're using lists within each wp_triggers_cache entry.

Reply:
Usually one tdata1 value corresponds to many tdata2 values, so this implementation seemed more optimal

Reply:
Can you give an example where you have multiple tdata2 values for a single trigger?

Reply:
I guess this happens when a single trigger is used for different watchpoints/breakpoints . Like user set a breakpoint at some address, then disabled it and set at some other address. This will result in accumulation of different `tdata2` values.

Reply:
But why do we care to accumulate different `tdata2` values for a single trigger? I thought the point was for OpenOCD to keep track of what `tdata2` is for each trigger, so it doesn't need to read it from the target, which is slow.

Reply:
As far as I understand the point is to cache non-successful results of calls to `set_trigger` (`ERROR_TARGET_RESOURCE_NOT_AVAILABLE`). This makes an assumption that if a combination of `tdata1` and `tdata2` resulted in `ERROR_TARGET_RESOURCE_NOT_AVAILABLE` once, the result will be the same. Such structure allows to skip attempts to set a trigger.
e.g.:
* A breakpoint needs to be restored.
* The required combination of `tdata1` and `tdata2` is only supported on the last trigger.
* While the breakpoint was set for the first time, unsuccessful attempts to set it on the triggers with lower indexes were collected into the cache.
* These attempts will be skipped during restore (no communication with DM is required).

Reply:
Ah, that's the part I was missing! Thanks for explaining. I'll review again.

Reply:
And of course the original description said "It significantly speeds things up, since we cache failed attempts". Sorry for not reading more closely.